src/object.jl

```diff
@@ -372,3 +372,135 @@ function Base.read(io::PrefetchedDownloadStream, ::Type{UInt8})
     end
     return b
 end
+
+function _upload_task(io; kw...)
+    try
+        (part_n, upload_buffer) = take!(io.upload_queue)
+        # upload the part
+        parteTag, wb = uploadPart(io.store, io.url, upload_buffer, part_n, io.uploadState; io.credentials, kw...)
+        # add part eTag to our collection of eTags in the right order
+        put!(io.sync, part_n) do
+            push!(io.eTags, parteTag)
+        end
+        Base.@lock io.cond_wait begin
+            io.ntasks -= 1
+            notify(io.cond_wait)
+        end
+    catch e
+        isopen(io.upload_queue) && close(io.upload_queue, e)
+        Base.@lock io.cond_wait begin
+            io.is_error = true
+            io.exc = e
+            notify(io.cond_wait, e, all=true, error=true)
+        end
+    end
+    return nothing
+end
+
+"""
+    MultipartUploadStream <: IO
+    MultipartUploadStream(args...; kwargs...) -> MultipartUploadStream
+
+An in-memory IO stream that uploads chunks to a URL in blob storage.
+
+Data chunks are written to a channel and spawned tasks read buffers from this channel.
+Each task uploads a distinct part to blob storage.
+Each part is tagged with an id and we keep the tags into a vector. We also increment a
+counter to keep track of the number of parts. When there are no more data chunks to upload,
+we send a POST request with a single id for the entire upload.
+
+# Arguments
+* `store::AbstractStore`: The S3 Bucket / Azure Container object
+* `key::String`: S3 key / Azure blob resource name
+
+# Keywords
+* `credentials::Union{CloudCredentials, Nothing}=nothing`: Credentials object used in HTTP
+    requests
+* `kwargs...`: HTTP keyword arguments are forwarded to underlying HTTP requests,
+
+## Examples
+```
+# Get an IO stream for a remote CSV file `test.csv` living in your S3 bucket
+io = MultipartUploadStream(bucket, "test.csv"; credentials)
+```
+"""
+mutable struct MultipartUploadStream <: IO
+    store::AbstractStore
+    url::String
+    credentials::Union{Nothing, AWS.Credentials, Azure.Credentials}
+    uploadState
```

Comment:
Should we specialize on the `store` type? What is the type of `uploadState`? Ideally, we'd specialize so that touching the credentials and store is inferrable

Reply:
I replaced `AbstractStore`  for `store` with `Union{AWS.Bucket, Azure.Container}`.  `uploadState` is either `String` or `nothing`.

Reply:
But do we want specialize MultipartUploadStream{T<:AbstractStore}, similar to `PrefetchedDownloadStream`? It's possible that Julia will manage to union split on small unions like these but it would be nice to make sure we don't box unnecessarily. Also to enforce that we have compatible Credetial and Store objects (i.e. both Azure or both AWS specific).

Reply:
I don't mind making `MultipartUploadStream` being parametrized on `AbstractStore`, but I'm not quite following what's the benefit in comparison to `Union{AWS.Bucket, Azure.Container}`, as I'm not very familiar with the details of boxing. We do use a union for credentials. Is this because there isn't an abstract struct for those?

Reply:
Maybe boxing is not the right CS term, but the idea is to make the code as inferrable as possible, so Julia can optimize things easily without having to do dynamic dispatch and so on. In these type-unstable cases, Julia tends to allocate defensively because it doesn't know which types it will encounter. In this specific example, it might be ok since the unions are small, but there is no harm in specializing these since people tend to know in advance if they want to talk to Azure or S3 anyway.

Reply:
Thanks for the clarification! I admit I like the flexibility of `Union` 😄, but I see your point.